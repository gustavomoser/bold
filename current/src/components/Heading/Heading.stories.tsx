import { select } from '@storybook/addon-knobs'
import React from 'react'

import { TextColor } from '../../styles'
import { VFlow } from '../VFlow'

import { Heading } from './Heading'

const colors: TextColor[] = [
  '' as any,
  'inherit',
  'normal',
  'secondary',
  'disabled',
  'primary',
  'danger',
  'info',
  'alert',
  'success',
]

export default {
  title: 'Components/Textual',
}

export const _Heading = () => (
  <VFlow>
    <Heading level={1} color={select('color', colors, 'normal')}>
      Heading level 1
    </Heading>
    <Heading level={2} color={select('color', colors, 'normal')}>
      Heading level 2
    </Heading>
    <Heading level={3} color={select('color', colors, 'normal')}>
      Heading level 3
    </Heading>
    <Heading level={4} color={select('color', colors, 'normal')}>
      Heading level 4
    </Heading>
    <Heading level={5} color={select('color', colors, 'normal')}>
      Heading level 5
    </Heading>
    <Heading level={6} color={select('color', colors, 'normal')}>
      Heading level 6
    </Heading>
  </VFlow>
)
