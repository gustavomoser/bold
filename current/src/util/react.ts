import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
