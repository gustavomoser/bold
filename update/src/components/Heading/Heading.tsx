import React from 'react'

import { Text, TextProps } from '../Text'

export interface HeadingProps extends TextProps {
  level: 1 | 2 | 3 | 4 | 5 | 6
}

export function Heading(props: HeadingProps) {
  const { level, ...rest } = props

  const heading = `h${level}` as 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'

  return <Text component={heading} variant={heading} {...rest} />
}
