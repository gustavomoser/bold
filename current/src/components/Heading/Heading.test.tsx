import { render } from '@testing-library/react'
import React from 'react'

import { Heading } from './Heading'

it('should render correctly', () => {
  expect(render(<Heading level={1}>Heading 1</Heading>).container).toMatchSnapshot()
  expect(render(<Heading level={2}>Heading 2</Heading>).container).toMatchSnapshot()
  expect(render(<Heading level={3}>Heading 3</Heading>).container).toMatchSnapshot()
  expect(render(<Heading level={4}>Heading 4</Heading>).container).toMatchSnapshot()
  expect(render(<Heading level={5}>Heading 5</Heading>).container).toMatchSnapshot()
  expect(render(<Heading level={6}>Heading 6</Heading>).container).toMatchSnapshot()
})

it('should accept all text props', () => {
  const { container } = render(
    <Heading level={1} color='primary' fontWeight='bolder' fontSize={4} textDecoration='underline' fontStyle='italic'>
      Heading with all text props
    </Heading>
  )
  expect(container).toMatchSnapshot()
})

it('should accept style prop', () => {
  const { container } = render(
    <Heading level={1} style={{ color: 'red' }}>
      Heading Primary color
    </Heading>
  )
  expect(container).toMatchSnapshot()
})
