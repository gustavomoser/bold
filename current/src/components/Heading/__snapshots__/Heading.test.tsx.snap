// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept all text props 1`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 1.5rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
  color: #0069D0;
  font-size: 4rem;
  font-weight: bolder;
  font-style: italic;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div>
  <h1
    class="emotion-0"
  >
    Heading with all text props
  </h1>
</div>
`;

exports[`should accept style prop 1`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 1.5rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
  color: red;
}

<div>
  <h1
    class="emotion-0"
  >
    Heading Primary color
  </h1>
</div>
`;

exports[`should render correctly 1`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 1.5rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
}

<div>
  <h1
    class="emotion-0"
  >
    Heading 1
  </h1>
</div>
`;

exports[`should render correctly 2`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 1.25rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
}

<div>
  <h2
    class="emotion-0"
  >
    Heading 2
  </h2>
</div>
`;

exports[`should render correctly 3`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 1rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
}

<div>
  <h3
    class="emotion-0"
  >
    Heading 3
  </h3>
</div>
`;

exports[`should render correctly 4`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.875rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
}

<div>
  <h4
    class="emotion-0"
  >
    Heading 4
  </h4>
</div>
`;

exports[`should render correctly 5`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
}

<div>
  <h5
    class="emotion-0"
  >
    Heading 5
  </h5>
</div>
`;

exports[`should render correctly 6`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.75rem;
  line-height: 1.5;
  color: #24252E;
  font-weight: bold;
}

<div>
  <h6
    class="emotion-0"
  >
    Heading 6
  </h6>
</div>
`;
